File: Python/Graph/1443_Minimum Time to Collect All Apples in a Tree (Microsoft).py
class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        graph = collections.defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        
        def dfs(graph, hasApple, node, parent):
            totalTime = 0
            childTime = 0
            
            for child in graph[node]:
                if child == parent:
                    continue
                childTime = dfs(graph, hasApple, child, node)
                    
                if(childTime > 0 or hasApple[child]):
                    totalTime += 2+ childTime
                
            return totalTime
        
        
        return dfs(graph, hasApple, 0, -1)
                    

File: C++/Graph/1443_Minimum Time to Collect All Apples in a Tree (Microsoft).cpp
class Solution {
public:
    
    int dfs(unordered_map<int, vector<int>> &adj, vector<bool>& hasApple, int node, int parent){
        int totaltime = 0; int childtime = 0;
        
        
        for(auto child:adj[node]){
            if(child==parent)
                continue;
            childtime = dfs(adj, hasApple, child, node);
            if(childtime > 0 || hasApple[child])
                totaltime += childtime + 2;  
            
            
        }
        
        
        
        return totaltime;
        
    }
    
    
    
    
    int minTime(int n, vector<vector<int>>& edges, vector<bool>& hasApple) {
        unordered_map<int, vector<int>> adj;
        for(auto vec:edges){
            adj[vec[0]].push_back(vec[1]);
            adj[vec[1]].push_back(vec[0]);            
        }
        
        return dfs(adj, hasApple,0,-1);
    }
    
    
    
    
};